contracts/javascore/ibc/src/main/java/ibc/ics20/app/ICS20Transfer.java

```diff
@@ -0,0 +1,179 @@
+//        TODO unmarshal json
+        boolean success = false;
+        Address receiver = ZERO_ADDRESS;
+        receiver, success = _decodeReceiver(packet.data.receiver);
+        if (!success) {
+           return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+
+        byte[] denomPrefix = getDenomPrefix(packet.data.sourcePort, packet.data.sourceChannel);
+        byte[] denom = packet.data.denom.getBytes();
+        if (denom.length >= denomPrefix.length) {
+//            denom slicing todo
+            success = _transferFrom(getEscrowAddress(packet.data.sourceChannel), receiver, packet.data.denom, packet.data.amount);
+        } else{
+            if(ICS20Lib.isEscapeNeededString(denom)){
+                success = false;
+            }else{
+                success = _mint(receiver, String(getDenomPrefix(packet.destination_port, packet.destination_channel), denom), packet.data.amount);
+            }
+        }
+
+        if (success) {
+            return ICS20Lib.SUCCESSFUL_ACKNOWLEDGEMENT_JSON;
+        }else{
+            return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+
+
+        return packet;
+    }
+
+
+    @External
+    public void onAcknowledgementPacket(byte[] calldata, byte[] acknowledgement, Address relayer) {
+        Context.println("onAcknowledgementPacket");
+//        Context.require(Context.getCaller().equals(getIBCAddress()), "caller is not handler");
+        if (acknowledgement != ICS20Lib.SUCCESSFUL_ACKNOWLEDGEMENT_JSON) {
+                refundTokens(calldata, packet.source_port, packet.source_channel);
+        }
+
+    }
+
+    @External
+    public String onChanOpenInit(IIBCModule.onChanOpenInit msg) {
+//        srcChan.set(channelId);
+//        srcPort.set(portId);
+//        Channel.Counterparty counterparty = Channel.Counterparty.decode(counterpartyPb);
+//        dstPort.set(counterparty.getPortId());
+//        Context.println("onChanOpenInit");
+        return ICS20_VERSION;
+    }
+
+    @External
+    public String onChanOpenTry(IIBCModule.onChanOpenTry msg) {
+        return ICS20_VERSION;
+    }
+
+    @External
+    public void onChanOpenAck(IIBCModule.MsgOnChanOpenAck msg) {
+
+    }
+
+    @External
+    public void onChanCloseInit(IIBCModule.MsgOnChanCloseInit msg) {
+        Context.revert("Not Allowed");
+    }
+
+    @External
+    public void onTimeoutPacket(Packet packet, byte[] proofHeight, byte[] proof, BigInteger nextSequenceRecv) {
+    }
+
+
+    static Address getEscrowAddress(String sourceChannel) {
+        Address escorw = channelEscrowAddresses.get(sourceChannel);
+        Context.require(escorw != ZERO_ADDRESS);
+        return escorw;
+    }
+
+    private void refundTokens(ICS20Lib.PacketData data, String sourcePort, String sourceChannel) {
+        byte[] denomPrefix = getDenomPrefix(sourcePort, sourceChannel);
+        byte[] denom = data.denom.getBytes();
+
+        if (denom.length >= denomPrefix.length) {
+
```

Comment:
this should be called on both on ack error and timeout and in both cases packet will be the one we sent . so based on packets destinationPort /destinationChannel and denom we can determine if receiving end is source or not. Then we can proceed as required.

Reply:
resolved